docs/developer/architecture/centralized-cli.rst

```diff
@@ -149,3 +180,233 @@ Cluster Service Administration
+
+Input and Output
+""""""""""""""""
+
+Each command should follow the kubectl style and standard as much as possible:
+
+- Command style:
+
+  .. code-block:: bash
+
+    kubectl metalk8s <action> <resource>
+
+- Interactive:
+
+  No interaction with the user, except when it's an ``edit`` command an editor
+  is opened (if needed) and when a password is required a prompt appears to ask
+  it.
+
+- Output style:
+
+  Default human-readable output (``<object> <action>ed``).
+
+  A ``--output``, ``-o`` option to change output format, at least support
+  for ``json`` and ``yaml`` (``jsonpath`` and ``go-template`` when it's
+  possible).
+
+  Internally each action result should be an "object"
+  (e.g.: single-level dictionary) containing several informations,
+  at least:
+
+  - name
+  - message
+  - result (True or False)
+  - an elapsed time (to know each action time)
+
+  A ``--verbose``, ``-v`` option to change log level verbosity (default output
+  to ``stderr``), using
+  `Kubernetes log library <https://github.com/kubernetes/klog>`_.
+
+  By default each command will wait for a result but, when it's possible, a
+  ``--async`` option should allow to do not wait for a result and just output
+  an ID (e.g.: Job ID for Salt) that can be used to watch for the result.
+
+SaltAPI interaction
+"""""""""""""""""""
+
+If the plugin needs to access Salt API then it should use the service proxy
+``http://<apiserver_host>/api/v1/namespaces/kube-system/services/https:salt-master:api/proxy/``.
+
+For each and every Salt API call plugin will need authentication on
+apiserver to access the Salt API service proxy and also to Salt API.
+
+.. note::
+
+  Right now, Salt API only accepts authentication using Bearer token, but in
+  kubeconfig we could have certificates authentication so this kind of
+  kubeconfig will not work with this kubectl plugin.
+
+  Add support for certificates based authentication in Salt API look quite hard
+  and costly.
+
+.. todo::
+
+  Check what can be done to support certificates based kubeconfig or if we
+  consider that we only support token-based kubeconfig for our plugin commands
+  interacting with Salt API.
+
+Deployment
+""""""""""
+
+Plugin should be developed as one single binary ``kubectl-metalk8s`` available
+from the ISO, easy buildable from GitHub repository and also as a System
+Package for Operating System supported by MetalK8s.
+
+The package should install the plugin in ``/usr/bin`` directory by default.
+
+This package should be installed on the bootstrap node by default.
+
+Rejected design choice
+""""""""""""""""""""""
+
+- ``Bash`` kubectl plugin: Bash is great to do simple actions but not when
+  you need to do interaction with some API like Kubernetes API or Salt API.
+- ``Python`` kubectl plugin: Python allows us to do complicated actions and
+  great interactions with APIs but interactions between ``Go`` and Kubernetes
+  are much better, with a sample CLI available.
```

Comment:
I'm not sure I understand what you mean by `interactions between ``Go`` and Kubernetes
  are much better, with a sample CLI available.` :thinking:

Reply:
`much simpler` maybe as you have several samples and sample CLI it's just another point for `Go` vs `Python` just the fact that you have a "Sample CLI plugin" from Kubernetes written in Go

Reply:
Ok I see.
That doesn't make the interaction easier per se, but indeed make the development easier because you have lot of example around.

Maybe:

> but interactions between ``Go`` and Kubernetes are much easier, given the large number of example available.